src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/NullableHelpers/NullableHelpers.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static class NullableHelpers
+    {
+        /// <summary>
+        /// Gets the declared symbol and root operation from the passed in declarationSyntax and calls
+        /// <see cref="IsSymbolAssignedPossiblyNullValue(SemanticModel, IOperation, ISymbol)"/>. Note
+        /// that this is bool and not bool? because we know that the symbol is at the very least declared, 
+        /// so there's no need to return a null value. 
+        /// </summary>
+        public static bool IsDeclaredSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, SyntaxNode declarationSyntax, CancellationToken cancellationToken)
+        {
+            var declaredType = semanticModel.GetRequiredDeclaredSymbol(declarationSyntax, cancellationToken);
+            var declaredOperation = semanticModel.GetRequiredOperation(declarationSyntax, cancellationToken);
+
+            var rootOperation = declaredOperation;
+
+            // Walk up the tree to find a root for the operation
+            // that contains the declaration
+            while (rootOperation is not IBlockOperation &&
+                rootOperation.Parent is not null)
+            {
+                rootOperation = rootOperation.Parent;
+            }
+
+            return IsSymbolAssignedPossiblyNullValue(semanticModel, rootOperation, declaredType) == true;
+        }
+
+        /// <summary>
+        /// Given an operation, goes through all decendent operations and returns true if the symbol passed in
+        /// is ever assigned a possibly null value as determined by nullable flow state. Returns
+        /// null if no references are found, letting the caller determine what to do with that information
+        /// </summary>
+        public static bool? IsSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, IOperation operation, ISymbol symbol)
+        {
+            var references = operation.DescendantsAndSelf()
+                .Where(o => IsSymbolReferencedByOperation(o, symbol));
+
+            var hasReference = false;
+
+            foreach (var reference in references)
+            {
+                hasReference = true;
+
+                // foreach statements are handled special because the iterator is not assignable, so the elementtype 
+                // annotation is accurate for determining if the loop declaration has a reference that allows the symbol
+                // to be null
```

Comment:
i don't really get what hte comment means.  don't we still care how the varaible is used to dtermine if we should really think of it as `T?` vs `T`?

Reply:
What do you mean by used? `T?` and `T` are only relevant for assignment. Null flow analysis won't change beyond the declared capability here unless it somehow considers enumerable constant values.